cmake/modules/extensions.cmake

```diff
@@ -5166,62 +5165,84 @@ function(add_llext_target target_name)
       "$<FILTER:${zephyr_flags},EXCLUDE,${llext_remove_flags_regexp}>"
   )
 
-  # Compile the source file to an object file using current Zephyr settings
-  # but a different set of flags
-  add_library(${target_name}_lib OBJECT ${source_file})
-  target_compile_definitions(${target_name}_lib PRIVATE
+  # Compile the source file using current Zephyr settings but a different
+  # set of flags.
+  # This is currently arch-specific since the ARM loader for .llext files
+  # expects object file format, while the Xtensa one uses shared libraries.
+  set(llext_lib_target ${target_name}_llext_lib)
+  if(CONFIG_ARM)
```

Comment:
> then having (too much) CONFIG_<feature> is a bad sign.

It is not good but "spreading" is IMHO slightly exaggerated and @pillo79 has committed above to fix this as soon as the corresponding C code is fixed - if nothing else for the simple reason that it makes his own maintenance work harder.

The LLEXT paint is still wet and we need more early adopters (like SOF) to starting using LLEXT in anger and provide feedback to make progress. One thing slowing down adoption and progress has been (as often) the lack of a build system, forcing everyone to write their own, different, custom build script. So @pillo79 stepped up and spent an unbelievable amount of time and patience addressing that gap in the CMake area where extremely few people ever volunteer. Blocking the (extension of) this small, documented CMake hack can re-inforce a chicken-and-egg situation where the corresponding C code lacks users and does not get fixed either.

From experience I know very well that "Fix it later" usually means either "never" or "someone else can do it" but I think this case deserves an exception  because:
- LLEXT is still at the (late) prototype stage so there's barely any user base and nothing "stable" to break yet.
- It's not even a new hack, it's already there
- @pillo79 's dedication is a good indication we can trust him to actually fix this later
- we have no one else :-D

Reply:
> It is not good but "spreading" is IMHO slightly exaggerated

So which better word would you have in mind to describe the fact that the `if(CONFIG_...)` was used in one location before, and now two more locations are added ?
Or should I not point to the fact that such testing is not desired ?

Also, pointing at this could even have @pillo79 taking a second look at the code to see if this could be combined under the existing single `if(CONFIG_...)` test.

Remember, others may even look at this code and think this is an acceptable way, but for such cases it's always good to have review comments to reference and stating, `yes, but his pattern was accepted under specific conditions, see PR xyz`.

> The LLEXT paint is still wet and we need more early adopters (like SOF) to starting using LLEXT in anger and provide feedback to make progress. 

and I fully agree, and also a reason why I very early suggested that `llext` functions were made first class citizen and placed in `extensions.cmake`. But with that we should also require a bit higher level compared to code in a Zephyr module 🙂 

> @pillo79 's dedication is a good indication we can trust him to actually fix this later

and did I anywhere indicate that this was not the case ?

Reply:
The fact that something is observed in a PR doesn't mean that it cannot be accepted if reasonable explanation / answers are given.

Reply:
The "spreading" is due to the fact that build stages are now clearly separated, thus there are multiple (but closely packed) locations in which I need arch-specific behavior. I could probably minimize it somehow, but I chose readability over minimum code.

We can and probably will get rid of the arch-specific tags by defining behavior-specific Kconfigs (for example an enum like `CONFIG_LLEXT_TYPE_DYNAMIC`/`OBJECT`). But this is definitely out of scope of this PR, which is _not_ changing the "hack", just moving it around _to make it more explicit_.